File: src/main/java/dev/bnacar/distributedratelimiter/ratelimit/LeakyBucket.java
package dev.bnacar.distributedratelimiter.ratelimit;

import java.util.concurrent.BlockingQueue;
import java.util.concurrent.LinkedBlockingQueue;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.ScheduledExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.atomic.AtomicLong;

/**
 * Leaky bucket rate limiter implementation.
 * Enforces constant output rate through request queuing and processing.
 * Provides traffic shaping with predictable processing rates regardless of input bursts.
 */
public class LeakyBucket implements RateLimiter {
    
    private final int queueCapacity;
    private final double leakRatePerSecond;
    private final long maxQueueTimeMs;
    private final BlockingQueue<QueuedRequest> requestQueue;
    private final ScheduledExecutorService leakExecutor;
    private final AtomicLong lastLeakTime;
    private volatile boolean shutdown = false;
    
    /**
     * Represents a queued request waiting to be processed.
     */
    private static class QueuedRequest {
        final CompletableFuture<Boolean> future;
        final long enqueuedTime;
        final int tokens;
        
        QueuedRequest(int tokens) {
            this.future = new CompletableFuture<>();
            this.enqueuedTime = System.currentTimeMillis();
            this.tokens = tokens;
        }
    }
    
    /**
     * Create a new leaky bucket with specified parameters.
     * 
     * @param queueCapacity Maximum number of requests that can be queued
     * @param leakRatePerSecond Rate at which requests are processed (tokens per second)
     */
    public LeakyBucket(int queueCapacity, double leakRatePerSecond) {
        this(queueCapacity, leakRatePerSecond, 5000); // Default 5 second max queue time
    }
    
    /**
     * Create a new leaky bucket with specified parameters.
     * 
     * @param queueCapacity Maximum number of requests that can be queued
     * @param leakRatePerSecond Rate at which requests are processed (tokens per second)
     * @param maxQueueTimeMs Maximum time a request can wait in queue before timing out
     */
    public LeakyBucket(int queueCapacity, double leakRatePerSecond, long maxQueueTimeMs) {
        this.queueCapacity = queueCapacity;
        this.leakRatePerSecond = leakRatePerSecond;
        this.maxQueueTimeMs = maxQueueTimeMs;
        this.requestQueue = new LinkedBlockingQueue<>();
        this.lastLeakTime = new AtomicLong(System.currentTimeMillis());
        
        // Create dedicated executor for leak processing
        this.leakExecutor = Executors.newSingleThreadScheduledExecutor(r -> {
            Thread t = new Thread(r, "LeakyBucket-Leak");
            t.setDaemon(true);
            return t;
        });
        
        // Start leak processing - process every 100ms for smooth operation
        startLeakProcessing();
    }
    
    @Override
    public boolean tryConsume(int tokens) {
        if (tokens <= 0 || shutdown) {
            return false;
        }
        
        // Check if queue has capacity
        if (requestQueue.size() >= queueCapacity) {
            return false; // Queue full, reject immediately
        }
        
        // For synchronous operation, we need to simulate the queue behavior
        // In a real leaky bucket, this would be asynchronous, but the RateLimiter
        // interface is synchronous, so we calculate if the request would be processed
        
        long currentTime = System.currentTimeMillis();
        long estimatedProcessingTime = calculateEstimatedProcessingTime();
        
        // If estimated processing time exceeds max queue time, reject
        if (estimatedProcessingTime > maxQueueTimeMs) {
            return false;
        }
        
        // Add to queue (simulated - we don't actually queue for synchronous interface)
        return true;
    }
    
    /**
     * Calculate estimated time to process current queue plus new request.
     */
    private long calculateEstimatedProcessingTime() {
        int currentQueueSize = requestQueue.size();
        double processingTimePerToken = 1000.0 / leakRatePerSecond; // ms per token
        return (long) (currentQueueSize * processingTimePerToken);
    }
    
    @Override
    public int getCurrentTokens() {
        // For leaky bucket, return available queue capacity
        return Math.max(0, queueCapacity - requestQueue.size());
    }
    
    @Override
    public int getCapacity() {
        return queueCapacity;
    }
    
    @Override
    public int getRefillRate() {
        // Return leak rate as integer for interface compatibility
        return (int) Math.ceil(leakRatePerSecond);
    }
    
    @Override
    public long getLastRefillTime() {
        return lastLeakTime.get();
    }
    
    /**
     * Get the leak rate in requests per second.
     */
    public double getLeakRatePerSecond() {
        return leakRatePerSecond;
    }
    
    /**
     * Get the maximum queue wait time in milliseconds.
     */
    public long getMaxQueueTimeMs() {
        return maxQueueTimeMs;
    }
    
    /**
     * Get the current queue size.
     */
    public int getQueueSize() {
        return requestQueue.size();
    }
    
    /**
     * Start the leak processing task.
     */
    private void startLeakProcessing() {
        // Calculate interval based on leak rate for smooth processing
        long intervalMs = Math.max(10, (long) (1000 / leakRatePerSecond / 10)); // Process 10x per token interval
        
        leakExecutor.scheduleWithFixedDelay(
            this::processLeakage,
            0,
            intervalMs,
            TimeUnit.MILLISECONDS
        );
        
        // Also schedule timeout cleanup
        leakExecutor.scheduleWithFixedDelay(
            this::cleanupTimedOutRequests,
            1000, // Start after 1 second
            1000, // Run every second
            TimeUnit.MILLISECONDS
        );
    }
    
    /**
     * Process queued requests at the configured leak rate.
     */
    private void processLeakage() {
        if (shutdown || requestQueue.isEmpty()) {
            return;
        }
        
        long currentTime = System.currentTimeMillis();
        long lastLeak = lastLeakTime.get();
        long timeSinceLastLeak = currentTime - lastLeak;
        
        // Calculate how many tokens we can process based on elapsed time
        double tokensToProcess = (timeSinceLastLeak / 1000.0) * leakRatePerSecond;
        
        if (tokensToProcess >= 1.0) {
            int tokensProcessed = 0;
            int maxTokens = (int) Math.floor(tokensToProcess);
            
            // Process requests from queue
            while (tokensProcessed < maxTokens && !requestQueue.isEmpty()) {
                QueuedRequest request = requestQueue.poll();
                if (request != null) {
                    tokensProcessed += request.tokens;
                    request.future.complete(true);
                }
            }
            
            if (tokensProcessed > 0) {
                lastLeakTime.set(currentTime);
            }
        }
    }
    
    /**
     * Clean up requests that have been in the queue too long.
     */
    private void cleanupTimedOutRequests() {
        if (shutdown) {
            return;
        }
        
        long currentTime = System.currentTimeMillis();
        
        // Check for timed out requests at the front of the queue
        while (!requestQueue.isEmpty()) {
            QueuedRequest request = requestQueue.peek();
            if (request != null && (currentTime - request.enqueuedTime) > maxQueueTimeMs) {
                requestQueue.poll();
                request.future.complete(false); // Timeout
            } else {
                break; // Queue is ordered by time, so we can stop here
            }
        }
    }
    
    /**
     * Shutdown the leaky bucket and clean up resources.
     */
    public void shutdown() {
        shutdown = true;
        leakExecutor.shutdown();
        
        // Complete any remaining requests with false
        while (!requestQueue.isEmpty()) {
            QueuedRequest request = requestQueue.poll();
            if (request != null) {
                request.future.complete(false);
            }
        }
        
        try {
            if (!leakExecutor.awaitTermination(1, TimeUnit.SECONDS)) {
                leakExecutor.shutdownNow();
            }
        } catch (InterruptedException e) {
            leakExecutor.shutdownNow();
            Thread.currentThread().interrupt();
        }
    }
    
    /**
     * Asynchronously enqueue a request for processing.
     * This is the intended way to use the leaky bucket for true asynchronous operation.
     * 
     * @param tokens Number of tokens to consume
     * @return CompletableFuture that completes when the request is processed or times out
     */
    public CompletableFuture<Boolean> enqueueRequest(int tokens) {
        if (tokens <= 0 || shutdown) {
            return CompletableFuture.completedFuture(false);
        }
        
        // Check queue capacity
        if (requestQueue.size() >= queueCapacity) {
            return CompletableFuture.completedFuture(false);
        }
        
        QueuedRequest request = new QueuedRequest(tokens);
        if (requestQueue.offer(request)) {
            return request.future;
        } else {
            return CompletableFuture.completedFuture(false);
        }
    }
}

File: src/main/resources/scripts/leaky-bucket.lua
-- Redis Lua script for atomic leaky bucket operations
-- KEYS[1]: bucket key
-- ARGV[1]: queue capacity (max requests that can be queued)
-- ARGV[2]: leak rate per second (requests processed per second)
-- ARGV[3]: tokens to consume (usually 1 for request-based limiting)
-- ARGV[4]: current time in milliseconds
-- ARGV[5]: max queue time in milliseconds (timeout for queued requests)

local queue_key = KEYS[1]
local capacity = tonumber(ARGV[1])
local leak_rate = tonumber(ARGV[2])
local tokens_to_consume = tonumber(ARGV[3])
local current_time = tonumber(ARGV[4])
local max_queue_time = tonumber(ARGV[5])

-- Ensure we have valid inputs
if capacity <= 0 or leak_rate <= 0 or tokens_to_consume < 0 or max_queue_time <= 0 then
    return {-1, -1, capacity, leak_rate, current_time, 0}
end

-- Keys for queue and metadata
local metadata_key = queue_key .. ':meta'
local queue_list_key = queue_key .. ':queue'

-- Get current metadata
local metadata = redis.call('HMGET', metadata_key, 'last_leak_time', 'queue_size')
local last_leak_time = metadata[1]
local queue_size = metadata[2]

-- Initialize if this is the first access
if last_leak_time == false or last_leak_time == nil then
    last_leak_time = current_time
else
    last_leak_time = tonumber(last_leak_time)
end

if queue_size == false or queue_size == nil then
    queue_size = 0
else
    queue_size = tonumber(queue_size)
end

-- Clean up expired requests from the front of the queue
local cleanup_count = 0
while queue_size > 0 do
    -- Peek at the oldest request in the queue
    local oldest_request = redis.call('LINDEX', queue_list_key, 0)
    if oldest_request then
        local oldest_time = tonumber(oldest_request)
        if oldest_time and (current_time - oldest_time) > max_queue_time then
            -- Remove expired request
            redis.call('LPOP', queue_list_key)
            queue_size = queue_size - 1
            cleanup_count = cleanup_count + 1
        else
            break -- No more expired requests
        end
    else
        break
    end
end

-- Process requests at leak rate (simulate draining)
local time_elapsed = math.max(0, current_time - last_leak_time)
local tokens_to_process = math.floor(time_elapsed / 1000 * leak_rate)

if tokens_to_process > 0 and queue_size > 0 then
    -- Process (remove) requests from the front of the queue
    local processed_count = math.min(tokens_to_process, queue_size)
    for i = 1, processed_count do
        redis.call('LPOP', queue_list_key)
    end
    queue_size = queue_size - processed_count
    last_leak_time = current_time
end

-- Handle new request
local success = 0
local estimated_wait = 0

if tokens_to_consume == 0 then
    -- This is a query for current state, no consumption
    success = 0
elseif tokens_to_consume > 0 then
    -- Check if queue has capacity
    if queue_size + tokens_to_consume <= capacity then
        -- Add request(s) to queue
        for i = 1, tokens_to_consume do
            redis.call('RPUSH', queue_list_key, current_time)
        end
        queue_size = queue_size + tokens_to_consume
        success = 1
        
        -- Calculate estimated wait time
        estimated_wait = math.floor(queue_size / leak_rate * 1000) -- Convert to milliseconds
    else
        -- Queue full, reject
        success = 0
        estimated_wait = -1 -- Indicate rejection due to capacity
    end
end

-- Update metadata
redis.call('HMSET', metadata_key, 
    'last_leak_time', last_leak_time,
    'queue_size', queue_size,
    'capacity', capacity,
    'leak_rate', leak_rate)

-- Set expiration for cleanup after inactivity (24 hours)
redis.call('EXPIRE', metadata_key, 86400)
redis.call('EXPIRE', queue_list_key, 86400)

-- Return result: {success, queue_size, capacity, leak_rate, last_leak_time, estimated_wait_ms}
return {success, queue_size, capacity, leak_rate, last_leak_time, estimated_wait}